gaia-fast-list.js

```diff
@@ -866,6 +948,134 @@ Internal.prototype = {
   },
 
   /**
+   * Cache an image in memory so that
+   * is can be quickly retrieved next
+   * time the list-item needs to be
+   * rendered.
+   *
+   * If the user has set the length/size
+   * of the cache to falsy then we don't
+   * do any caching.
+   *
+   * TODO: Some list items may share the
+   * same image src. If we have a matching
+   * src in already in the cache we could
+   * reuse that object to save memory.
+   *
+   * Although this gets tricky when we want
+   * to revokeObjectURL() and don't know how
+   * many list-items depend on it.
+   *
+   * @param  {Number} index
+   * @param  {Blob|String} raw
+   * @return {Object}
+   * @private
+   */
+  cacheImage(image, index) {
+    if (!this.el.imageCacheLength) return;
+    if (!this.el.imageCacheSize) return;
+    this.images.hash[index] = image;
+    this.images.list.push(index);
+    this.images.bytes += image.bytes;
+    this.checkImageCacheLimit();
+    debug('cached image', image, this.images.bytes);
+    return image;
+  },
+
+  /**
+   * Attempt to fetch an image
+   * for the given item index.
+   *
+   * @param  {Number} index
+   * @return {Object|*}
+   * @private
+   */
+  getCachedImage(index) {
+    debug('get cached image', index);
+    return this.images.hash[index];
+  },
+
+  /**
+   * Check there is space remaining in
+   * the image cache discarding entries
+   * when we're full.
+   *
+   * The cache is full when either the
+   * size (bytes) or the length is breached.
+   * We use length as well as bytes as we
+   * don't want to store 1000s of strings
+   * in memory.
+   *
+   * When a limit is breached, we discard
+   * half of the cached images not currently
+   * in use, then we checkImageCacheLimit()
+   * once more.
+   *
+   * We aim to keep a relatively full cache
+   * to keep things fast, but don't want to
+   * be discarding every time a new image
+   * is loaded.
+   *
+   * @private
+   */
+  checkImageCacheLimit() {
+    var exceeded = this.images.bytes > this.el.imageCacheSize
+      || this.images.list.length > this.el.imageCacheLength;
+
+    if (!exceeded) return;
+
+    debug('image cache limit exceeded', exceeded);
+    var items = this.fastList.geometry.maxItemCount;
+    var images = this.images.list.length;
+    var discardCount = (images - items) / 2;
+    debug('discarding', discardCount);
+
+    while (discardCount-- > 0) {
+      this.discardOldestImage();
+      debug('bytes used', this.images.bytes);
+    }
+
+    // double-check we've freed enough memory
+    this.checkImageCacheLimit();
```

Comment:
if `imageCacheLength` = 12 but `maxItemCount` = 16 aren't we going to infinite loop?

it would definitely be a misconfiguration anyway but maybe we can hardened against this with a
`var items = Math.min(this.fastList.geometry.maxItemCount, this.el.imageCacheLength)` when `imageCacheLength` is defined.